libvuln/jsonblob/jsonblob.go

```diff
@@ -103,23 +121,94 @@ func (l *Loader) Err() error {
 	return l.err
 }
 
-// Store writes out the Store to the provided Writer. It's the inverse of Load.
+// Store writes out the contents of the receiver to the provided [io.Writer].
+// It's the inverse of [Load].
+//
+// Store may only be called once for a series of [Store.UpdateVulnerabilities] and
+// [Store.UpdateEnrichments] calls, as it deallocates resources as it writes them.
+//
+// It should be possible to call this as often as needed to flush resources to
+// disk.
 func (s *Store) Store(w io.Writer) error {
 	s.RLock()
 	defer s.RUnlock()
 	enc := json.NewEncoder(w)
 	enc.SetEscapeHTML(false)
-	for id, e := range s.entry {
-		for _, v := range e.Vuln {
-			if err := enc.Encode(&diskEntry{
-				CommonEntry: e.CommonEntry,
-				Ref:         id,
-				Vuln:        v,
-			}); err != nil {
-				return err
+	write := func(id uuid.UUID, e CommonEntry) func(driver.UpdateKind, *os.File, int) error {
+		return func(k driver.UpdateKind, f *os.File, ct int) error {
+			if f == nil {
+				return nil
+			}
+			defer f.Close()
+			shim := newBufShim(f)
+			defer shim.Close()
+			for i := 0; i < ct; i++ {
+				dent := diskEntry{
+					CommonEntry: e,
+					Ref:         id,
+					Kind:        k,
+				}
+				switch k {
+				case driver.EnrichmentKind:
+					dent.Enrichment = shim
+				case driver.VulnerabilityKind:
+					dent.Vuln = shim
+				default:
+					panic(fmt.Sprintf("programmer error: unknown kind: %v", k))
+				}
+				if err := enc.Encode(&dent); err != nil {
+					return err
+				}
 			}
+			return nil
 		}
 	}
+
+	for id, e := range s.entry {
+		w := write(id, e.CommonEntry)
+		verr := w(driver.VulnerabilityKind, e.vulns, e.vulnCt)
+		eerr := w(driver.EnrichmentKind, e.enrichments, e.enrichmentCt)
+		delete(s.entry, id)
+		if err := errors.Join(verr, eerr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BufShim treats every call to [bufShim.MarshalJSON] as a [bufio.Scanner.Scan]
+// call.
+//
+// Note this type is very weird, in that it can only be used as _either_ a
+// target or a source. Doing both on the same structure will silently do the
+// wrong thing.
```

Comment:
I don't follow this. Can you explain what is a `source` and a `target` in this context?

Reply:
Reworded